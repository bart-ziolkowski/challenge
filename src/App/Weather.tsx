import React, { FC, useEffect, useState } from "react";

interface WeatherProps {
  city: string | null;
}

const Weather: FC<WeatherProps> = ({ city }: WeatherProps) => {
  const [weatherInfo, setWeatherInfo] = useState({
    city: null as string | null,
    country: null as string | null,
    temp: null as number | null,
    icon: null as string | null,
    description: null as string | null,
  });

  const API_KEY = "4c4f0b1876954338598a7be96c66527b";
  //const API_KEY = process.env.REACT_APP_WEATHER_API_KEY;

  useEffect(() => {
    (async () => {
      if (city) {
        try {
          const geoResponse = await fetch(
            `https://api.openweathermap.org/geo/1.0/direct?q=${city}&appid=${API_KEY}`
          );
          const geoData = await geoResponse.json();

          const weatherResponse = await fetch(
            `https://api.openweathermap.org/data/2.5/weather?lat=${geoData[0].lat}&lon=${geoData[0].lon}&units=metric&appid=${API_KEY}`
          );
          const weatherData = await weatherResponse.json();

          setWeatherInfo({
            city: weatherData.name,
            country: weatherData.sys.country,
            temp: Math.floor(weatherData.main.temp),
            icon: `https://openweathermap.org/img/wn/${weatherData.weather[0].icon}@2x.png`,
            description: weatherData.weather[0].description,
          });
        } catch (error) {
          console.error("Error fetching weather data:", error);
        }
      }
    })();
  }, [city]);

  return (
    <>
      {weatherInfo.temp !== null && (
        <h1>
          {weatherInfo.city}, {weatherInfo.country}
        </h1>
      )}

      {weatherInfo.temp !== null && <p>{weatherInfo.temp} Celsius</p>}

      {weatherInfo.icon !== null && <img src={weatherInfo.icon} alt="Icon" />}

      {weatherInfo.description !== null && <p>{weatherInfo.description}</p>}
    </>
  );
};

export default Weather;
