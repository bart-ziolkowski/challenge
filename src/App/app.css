:root {
  font-size: 10px;
  font-family: "Segoe UI", Tahoma, Geneva, Verdana, sans-serif;
  --teal: #1fb8b8;
  --ruby: #950e45;
}

:root[data-theme="dark"] {
  --background: #1f1f1f;
  --text: #f1f1f1;
}

:root[data-theme="light"] {
  --background: #f1f1f1;
  --text: #1f1f1f;
}

:root[data-theme="light"] div {
  background-color: #fee569;
}

:root[data-theme="light"] button {
  color: var(--ruby);
}

:root[data-theme="light"] button:hover {
  color: var(--background);
}

html,
body,
* {
  box-sizing: border-box;

  margin: 0;
  padding: 0;
  border: none;

  font-size: 1.5rem;
  text-transform: none;
  text-decoration: none;
}

body {
  display: grid;
  place-items: center;
  min-height: 100vh;

  background-color: var(--background);
  color: var(--text);
}

div {
  display: grid;
  grid-template: repeat(3, min-content) / repeat(2, 25rem);
  grid-template-areas:
    "input input"
    "submit theme"
    "city temp"
    "icon icon"
    "description description";
  gap: 2rem;

  padding: 5rem;
  border: 2px solid var(--ruby);
  border-radius: 1rem;

  text-align: center;
}

input,
button {
  height: 5rem;
  border-radius: 1rem;
}

input {
  grid-area: input;
  padding-inline: 2rem;
}

button {
  grid-area: submit;
  border: 2px solid var(--teal);

  background-color: var(--background);
  color: var(--teal);

  cursor: pointer;
  transition: transform 0.15s ease-in-out, border-color 0.15s ease-in-out,
    color 0.15s ease-in-out;
}

button:hover,
button:focus {
  background-color: var(--text);
  color: var(--background);
  border-color: var(--text);

  transform: scale(1.05);
}

h1 {
  grid-area: city;
}

p {
  grid-area: temp;
}

p:nth-of-type(2) {
  grid-area: description;
  text-transform: capitalize;
}

img {
  grid-area: icon;
  justify-self: center;
  margin: -3rem;
  height: 10rem;
}

.theme-button {
  grid-area: theme;
}

@media (max-width: 670px) {
  div {
    grid-template: repeat(3, min-content) / 1fr;
    grid-template-areas:
      "input"
      "submit"
      "theme"
      "city"
      "temp"
      "icon"
      "description";
    margin: 1rem;
  }
}
