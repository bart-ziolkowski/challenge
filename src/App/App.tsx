import "./app.css";

import React, { FC, useEffect, useRef, useState } from "react";

import Weather from "./Weather";

interface AppProps {}

const App: FC<AppProps> = () => {
  const cityInputRef = useRef<HTMLInputElement | null>(null);
  const [city, setCity] = useState("");
  const [theme, setTheme] = useState("dark");

  const toggleTheme = () => {
    setTheme((prevTheme) => (prevTheme === "dark" ? "light" : "dark"));
  };

  useEffect(() => {
    document.documentElement.setAttribute("data-theme", theme);
  }, [theme]);

  return (
    <>
      <button className="theme-button" onClick={() => toggleTheme()}>
        {theme === "dark" ? "Light" : "Dark"} Mode
      </button>

      <input
        ref={cityInputRef}
        role="search"
        type="text"
        placeholder="Enter city name..."
      />

      <button onClick={() => setCity(cityInputRef.current?.value || "")}>
        Show Weather
      </button>

      <Weather city={city} />
    </>
  );
};

export default App;
