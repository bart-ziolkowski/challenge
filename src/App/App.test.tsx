import { render, screen } from "@testing-library/react";

import App from "./App";
import userEvent from "@testing-library/user-event";

describe("App", () => {
  it("Renders Component According To Snapshot", () => {
    const { container } = render(<App />);

    expect(container).toMatchSnapshot();
  });

  it("Checks Input Functionality", () => {
    render(<App />);

    userEvent.type(screen.getByRole("search"), "Copenhagen");
    expect(screen.getByRole("search")).toHaveValue("Copenhagen");
  });
});
