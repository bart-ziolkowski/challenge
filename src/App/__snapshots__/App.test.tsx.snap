// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Renders Component According To Snapshot 1`] = `
<div>
  <button
    class="theme-button"
  >
    Light
     Mode
  </button>
  <input
    placeholder="Enter city name..."
    role="search"
    type="text"
  />
  <button>
    Show Weather
  </button>
</div>
`;
